File: 0954-maximum-sum-circular-subarray/0954-maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        maxi = nums[0]
        mini = nums[0]
        curmax = 0
        curmin = 0
        sum = 0
        for i in nums:
            sum += i
            curmax = max(0,curmax)
            curmax += i
            maxi = max(curmax, maxi)

            curmin = min(0, curmin)
            curmin += i
            mini = min(curmin, mini)
        if sum != mini: 
            return max(maxi, sum - mini)
        return maxi

File: 0954-maximum-sum-circular-subarray/0954-maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int maxi = nums[0];
        int mini = nums[0];
        int curmax = 0;
        int curmin = 0;
        int sum = 0;
        for(int i = 0; i < nums.size(); i++){
            sum += nums[i];
            curmax = max(curmax, 0);
            curmin = min(curmin, 0);
            curmax += nums[i];
            curmin += nums[i];
            maxi = max(maxi, curmax);
            mini = min(mini, curmin);
        }
        if(sum == mini){
            return maxi;
        }
        return max(maxi, sum - mini);
    }
};